[wip] Add shim layer to abstract dependencies on mbed OS

salkinium opened the pull request:
This allows `utest` to be used outside of mbed OS.

There are two scheduler implementations in the shim:
- using MINAR for mbed OS (default),
- using `us_ticker` for mbed OS (and perhaps mbedmicro?)

If yotta is detected, MINAR will be used, else `us_ticker` is used.

To use `utest` outside of mbed OS you also need to provide a critical section implementation using these two macros:
- `UTEST_ENTER_CRITICAL_SECTION`, and
- `UTEST_LEAVE_CRITICAL_SECTION`.

The usage of the macros inside the Harness is compatible with RAII, so it is possible to use a RAII class for its purpose:

```
#define UTEST_ENTER_CRITICAL_SECTION YourRAIIClass lock
```

Note: This code has not been tested with mbedmicro.

cc @0xc0170 @PrzemekWirkus

niklarm committed:
Extend scheduler with init and run callbacks.

niklarm committed:
Adapt Harness to call init and run functions.

niklarm committed:
Adapt examples to use init and run callbacks.

niklarm committed:
Add shim definitions for mbed OS definitions.

This allows utest to be used without mbed OS dependencies.

niklarm committed:
Use scheduler provided by shim layer in Harness.

niklarm committed:
Use shim macros for critical sections in Harness.

PrzemekWirkus commented:
> If yotta is detected, MINAR will be used, else us_ticker is used.

Maybe assuming that using ```yotta``` will automatically enable ```MINAR``` is not the best idea.
On the other hand we should be backward compatible with previous implementations.

niklas-arm commented:
> Maybe assuming that using yotta will automatically enable MINAR is not the best idea.

It actually checks if `YOTTA_MINAR_VERSION_STRING` is defined, which it is only, if you included the module in yotta.

0xc0170 commented:
How to introduce custom scheduler into the utest code base? would utest_config header file be a good answer? Or to tweak a build configuration to get those macros in? If a config header file, how to make it backward compatible within this PR?

I went with utest config for a quick test of this new API, I created utest_config.h:

```
// just a test, don't take this seriously
#define UTEST_ENTER_CRITICAL_SECTION __disable_irq()
#define UTEST_LEAVE_CRITICAL_SECTION __enable_irq()

// custom scheduler
#define YOTTA_CFG_UTEST_USE_CUSTOM_SCHEDULER 1
```

I had to add ``#include utest_config.h`` to shim.h (= using external build system, which does not allow me to add macros), implemented my own custom scheduling via provided API, and was able to run it.

niklas-arm commented:
Hm, ok, so different idea without using macros: If you want to port utest to another platform, you need to implement these three functions:

- `void utest_enter_critical_section(void)`
- `void utest_leave_critical_section(void)`
- `utest_v1_scheduler_t utest_get_scheduler(void)`

Similar to how it is done now, there would be default implementations for when use of Minar and Core Util modules in yotta have been detected, otherwise the functions will have to be implemented and linked externally.

niklarm committed:
Use functions instead of macros for porting.

0xc0170 commented on utest/shim.h:
```diff
@@ -0,0 +1,81 @@
+#       define UTEST_SHIM_SCHEDULER_USE_MINAR UTEST_MINAR_AVAILABLE
+#   endif
+#   ifndef UTEST_SHIM_SCHEDULER_USE_US_TICKER
+#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER (!UTEST_MINAR_AVAILABLE)
```
This effectively enables us ticker if a user does not supply any of minar version string or use minar. Shouldn't we leave it (no macros defined for utest), and a user needs to to implement the required API?

niklas-arm replied:
Whooops

0xc0170 commented on source/shim.cpp:
```diff
@@ -0,0 +1,133 @@
+// only one callback is active at any given time
+static volatile utest_v1_harness_callback_t minimal_callback;
+static volatile utest_v1_harness_callback_t ticker_callback;
+static const ticker_data_t *ticker_data;
```
This should pull in the us ticker header file

niklas-arm replied:
Something like this? `#include "us_ticker.h"` is the correct path for mbed?
```c
#ifdef YOTTA_MBED_HAL_VERSION_STRING
#   include "mbed-hal/us_ticker.h"
#else
#   include "us_ticker.h"
#endif
```

0xc0170 replied:
Yes that should do it

niklarm committed:
Don't enable us_ticker scheduler if minar unavail.

0xc0170 commented on utest/shim.h:
```diff
@@ -59,7 +59,7 @@
 #   endif
 #   ifndef UTEST_SHIM_SCHEDULER_USE_US_TICKER
-#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER (!UTEST_MINAR_AVAILABLE)
+#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER 0
```
would it make sense to enable us ticker by default for mbed SDK (__MBED__=1) ?
```
#       if defined(__MBED__) && (__MBED__ == 1)
#           define UTEST_SHIM_SCHEDULER_USE_US_TICKER 1
#       else
#           define UTEST_SHIM_SCHEDULER_USE_US_TICKER 0
#       endif
```

niklas-arm replied:
I added you as a collaborator, so you can make the changes faster, as I cannot test this right now?

niklas-arm replied:
And to the original question: Yes, I think so.

niklarm committed:
Include us_ticker.h in scheduler implementation.

0xc0170 committed:
Shim - mbed SDK uses us ticker by default

mbed SDK defines ``__MBED__`` macro (unique, not valid for yotta),
this should turn on us ticker by default, as that one is available
for mbed SDK targets.

0xc0170 commented:
@niklas-arm I added us ticker by default for mbed SDK  targets. As it is now, worked for my mbed SDK use. I haven't tested this branch wtih yotta tests.

niklas-arm commented:
I just tested this and it works with yotta.

PrzemekWirkus commented:
> I just tested this and it works with yotta.

@niklas-arm How about it works with us_ticker ? Did you try this solution.

niklas-arm commented:
`us_ticker.h` was moved to `us_ticker_api.h`, that's why it broke, see https://github.com/ARMmbed/utest/commit/e39d77c38d25e50b4407643725b260cf2553af9d.

0xc0170 commented:
> us_ticker.h was moved to us_ticker_api.h, that's why it broke, see e39d77c.

it was always ``us_ticker_api.h`` ?https://github.com/ARMmbed/mbed-hal/blob/master/mbed-hal/us_ticker_api.h

niklarm commented:
> it was always us_ticker_api.h ?https://github.com/ARMmbed/mbed-hal/blob/master/mbed-hal/us_ticker_api.h

Huh? Did this ever compile? Nope. Classic case of "wasn't tested".
Fixed now.
